timewarp_gl/plugin.cpp

```diff
@@ -228,62 +226,40 @@ class timewarp_gl : public threadloop {
 			}
 		}
 		// Construct a basic perspective projection
-		ksAlgebra::ksMatrix4x4f_CreateProjectionFov( &basicProjection, 40.0f, 40.0f, 40.0f, 40.0f, 0.1f, 0.0f );
-
+		math_util::projection_fov( &basicProjection, 40.0f, 40.0f, 40.0f, 40.0f, 0.1f, 0.0f );
 		// This was just temporary.
 		free(tw_mesh_base_ptr);
 
 		return;
 	}
 
 	/* Calculate timewarm transform from projection matrix, view matrix, etc */
-	void CalculateTimeWarpTransform( ksAlgebra::ksMatrix4x4f * transform, const ksAlgebra::ksMatrix4x4f * renderProjectionMatrix,
-                                        const ksAlgebra::ksMatrix4x4f * renderViewMatrix, const ksAlgebra::ksMatrix4x4f * newViewMatrix )
+	void CalculateTimeWarpTransform( Eigen::Matrix4f * transform, const Eigen::Matrix4f * renderProjectionMatrix,
+                                        const Eigen::Matrix4f * renderViewMatrix, const Eigen::Matrix4f * newViewMatrix )
 	{
-		// Convert the projection matrix from [-1, 1] space to [0, 1] space.
-		const ksAlgebra::ksMatrix4x4f texCoordProjection =
-		{ {
-			{ 0.5f * renderProjectionMatrix->m[0][0],        0.0f,                                           0.0f,  0.0f },
-			{ 0.0f,                                          0.5f * renderProjectionMatrix->m[1][1],         0.0f,  0.0f },
-			{ 0.5f * renderProjectionMatrix->m[2][0] - 0.5f, 0.5f * renderProjectionMatrix->m[2][1] - 0.5f, -1.0f,  0.0f },
-			{ 0.0f,                                          0.0f,                                           0.0f,  1.0f }
-		} };
+		Eigen::Matrix4f texCoordProjection;
+		texCoordProjection <<  0.5f * (*renderProjectionMatrix)(0,0),         0.0f,                                          0.5f * (*renderProjectionMatrix)(0,2) - 0.5f,  0.0f ,
```

Comment:
Add comment reminding readers that Eigen's default is column-major, not row-major.

Reply:
@mhuzai the initializer here (using the `<<` syntax) is actually storage-independent. The order in which the initializers are fed into the `<<` stream represents the actual structure of the matrix.

Are you talking about the `(*renderProjectionMatrix)(i,j)` accessors? If so, those are row-major. I.e., the first argument is which row, and the second argument is which column. Quoted from Eigen documentation:

> The primary coefficient accessors and mutators in Eigen are the overloaded parenthesis operators. **For matrices, the row index is always passed first.**

Reply:
I guess what is confusing me is that in the commit diff, the old code has a different initialization structure. It is exactly the transpose of the new structure. So I just assumed that it was because of Eigen's ordering. But if that is not the case, then why is the order different now? Was it column-major before?

Reply:
The old ksAlgebra matrices were initialized in **storage order,** column-major. Eigen matrices are initialized storage-order-independent, in "true order". In addition, the old ksAlgebra matrices were accessed `[i][j]`, where i = column and j = row. Eigen matrices are accessed oppositely.